KAFKA-2581: Run some existing ducktape tests with SSL

rajinisivaram opened the pull request:
Parametrize console consumer sanity test, replication tests and benchmarks tests to run with both PLAINTEXT and SSL.

rajinisivaram committed:
KAFKA-2581: Run some existing ducktape tests with SSL-enabled clients and brokers

ewencp commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -210,7 +215,10 @@ object ConsumerPerformance {
     
-    val props = new Properties
+    val props = if (options.has(consumerConfigOpt))
+      Utils.loadProps(options.valueOf(consumerConfigOpt))
```
Should these be added before or after the command line options? If we load them before, the default values for options would override custom settings from the `consumer.config` file even if the command line parameters were not specified.

ijuma replied:
Interesting. Ideally, _specified_ command-line settings would override the properties, but it seems like that would be a bigger change. Perhaps this should behave as the `ConsoleConsumer` until we can overhaul this. It looks like the change here does that, right?

ewencp replied:
Agreed, ideally if they were specified they would override, but if unspecified it would use values from the file. Not sure how easy that is to get working with this options library. I could have sworn I checked one of the other tools and found it using the behavior I expected, but now can't find it. I don't think this is make-or-break. I was just raising it since the behavior is unintuitive to me and although the performance tools are less critical, we probably wouldn't want to change their behavior that significantly after releasing a version with support for the `consumer.config` option.

rajinisivaram replied:
I do completely agree. However, I think all the tools currently do it this way and so I have kept it consistent. I will raise another JIRA to fix all the tools.

ewencp commented on tests/kafkatest/services/console_consumer.py:
```diff
@@ -132,6 +137,8 @@ def start_cmd(self):
             " --consumer.config %(config_file)s" % args
 
+        if self.new_consumer:
+            cmd += " --new-consumer --bootstrap-server %s"  % self.kafka.bootstrap_servers(self.security_protocol)
```
Since we're adding support for the new consumer instead of assuming old consumer, should we move the `--zookeeper` option into an else clause here?

ewencp commented on tests/kafkatest/services/console_consumer.py:
```diff
@@ -158,6 +165,14 @@ def _worker(self, idx, node):
+        if self.security_protocol is None:
+            self.security_protocol = security_config.security_protocol
+            if self.security_protocol == SecurityConfig.SSL:
+                self.new_consumer = True
```
It's kind of weird to override the `new_consumer` setting here. It works, but it makes it hard to reason about what's going on since the setting changes while starting up and may contradict the user-specified setting.

I see how it can make sense to set the `new_consumer` flag if the test requests SSL, but if we might override the setting, then maybe the default value in the constructor should be `None` and we should only set it to `True` here if it was left as `None`. Otherwise, we could throw an exception indicating they specified invalid combinations of settings.

More generally, do we actually need to wait until this point to figure all this out? I see that `SecurityConfig` is using `node.account`, but can some of this logic be resolved in the constructor, which makes the entire class simpler to reason about since all settings will be computed upon construction?

ijuma replied:
Agreed that we should throw an exception if an invalid combination is specified as some people may not be aware that SSL is not supported in the old consumer.

ewencp commented on tests/kafkatest/services/kafka.py:
```diff
@@ -56,10 +58,12 @@ def start(self):
         self.logger.info(props_file)
         node.account.create_file("/mnt/kafka.properties", props_file)
+        # Create keystore and truststore for SSL endpoint
+        SecurityConfig(node.account, SecurityConfig.SSL)
```
Should this always be passed `SSL`? It's unclear here how this works for `PLAINTEXT`. Or does this just work because the config for Kafka currently always specifies both `SSL` and `PLAINTEXT` listeners?

rajinisivaram replied:
Yes, Kafka has listeners with both PLAINTEXT and SSL in all tests. Clients choose the protocol with which to connect, and hence test one of the protocols. I will change Kafka to use only one listener to avoid confusion.

ewencp commented on tests/kafkatest/services/kafka.py:
```diff
@@ -96,7 +100,7 @@ def stop_node(self, node, clean_shutdown=True):
     def clean_node(self, node):
         node.account.kill_process("kafka", clean_shutdown=False, allow_fail=True)
-        node.account.ssh("rm -rf /mnt/kafka-logs /mnt/kafka.properties /mnt/kafka.log /mnt/kafka.pid", allow_fail=False)
+        node.account.ssh("rm -rf /mnt/kafka-logs /mnt/kafka.properties /mnt/kafka.log /mnt/kafka.pid /mnt/ssl", allow_fail=False)
```
See other comments about how `SecurityConfig` is used and whether a mixin would make sense. Seems less than ideal that we have to scatter `/mnt/ssl` throughout all the `clean-node` methods.

ewencp commented on tests/kafkatest/services/templates/kafka.properties:
```diff
@@ -20,6 +20,8 @@ port=9092
 #host.name=localhost
 advertised.host.name={{ node.account.hostname }}
 #advertised.port=<port accessible by clients>
+listeners=PLAINTEXT://:9092,SSL://:9093
```
I think we're going to have to protect all of this SSL stuff in conditionals (both blocks of settings added in this patch). Otherwise the version upgrade tests in KAFKA-1888 will break since these settings are not backwards compatible.

We can of course address this in KAFKA-1888 (which has an existing patch already), but it seems more relevant to add in this patch since this is the one adding non-default settings.

Also, do we want to have cases where we _only_ have SSL? It seems like cases where we're explicitly testing SSL should only turn on SSL so we don't, e.g., accidentally pass by using `PLAINTEXT` by mistake.

ijuma replied:
Yes, it would be good to have cases where we only have SSL, I think. For benchmarks with multiple brokers, it would also be interesting to test the case where SSL is used for the producer/consumer, but not for inter-broker communication (we would need multiple listeners then).

ewencp replied:
@ijuma Are these suggested tests captured in JIRAs somewhere? I haven't tracked the SSL issues closely, but I think getting an initial patch for existing tests checked in is a good start, and then we (hopefully easily) add these other tests in a separate patch.

rajinisivaram replied:
In the current patch, end-to-end latency tests are run with (protocol=PLAINTEXT,interbroker_protocol=PLAINTEXT), (protocol=SSL,interbroker_protocol=PLAINTEXT) and (protocol=SSL,interbroker_protocol=SSL). All other performance tests are run with (PLAINTEXT, PLAINTEXT and (SSL,SSL). I will add (SSL, PLAINTEXT) as well to all of the performance tests since it is quite easy to do.

ijuma replied:
Yes, agreed that any new tests should be in a separate PR. We don't have a definite list of tests to be written, although that is something we need to do.

I was just trying to explain that there will be cases where we'd want to have a mixed SSL and non-SSL connections. We can handle those when the need arises, if it's easier.

ijuma replied:
I wrote the above before I saw @rajinisivaram's message. @rajinisivaram, one thing to keep in mind is that interbroker_protocol only makes a difference if there are multiple brokers. If the tests are only using a single broker, then switching that property is a bit pointless. In the case where we have multiple brokers, I agree with the test combination you suggested.

ewencp commented on tests/kafkatest/tests/benchmark_test.py:
```diff
@@ -26,16 +28,22 @@
 
 
-class Benchmark(KafkaTest):
+class Benchmark(Test):
```
Why did we need to switch this from inheriting from `KafkaTest` to `Test`? We'd really like to avoid redundant code. The vast majority of tests really should be implemented via `KafkaTest`. Do we just need to generalize `KafkaTest` a bit more?

rajinisivaram replied:
This was changed because Kafka is now started in these tests using the security protocol specified in the test - hence the creation of the Kafka service can no longer be in the constructor. Since the current KafkaTest is suitable for tests which dont set security protocol, it seemed reasonable to leave it as is.

ewencp commented on tests/kafkatest/tests/benchmark_test.py:
```diff
@@ -44,25 +52,35 @@ def __init__(self, test_context):
+            self.zk, interbroker_security_protocol=interbroker_security_protocol, topics=self.topics)
+        self.kafka.start()
+
+    @parametrize(acks=1, topic=TOPIC_REP_ONE, num_producers=1, message_size=DEFAULT_RECORD_SIZE, security_protocol='PLAINTEXT')
```
I think we can use default values on the method to avoid having so many parameters to all these `@parametrize` calls, e.g. make `test_producer_throughput` use `num_producers=1, message_size=DEFAULT_RECORD_SIZE, security_protocol='PLAINTEXT'` so the `@parametrize` calls only need to specify values they override.

ewencp commented on tests/kafkatest/utils/security_config.py:
```diff
@@ -0,0 +1,132 @@
+# limitations under the License.
+
+import os
+import sys
```
This import is not used.

ewencp commented on tests/kafkatest/utils/security_config.py:
```diff
@@ -0,0 +1,132 @@
+import tempfile
+
+
+class Keytool:
```
Let's stick to new-style classes. This should be `class Keytool(object)`.

ewencp commented on tests/kafkatest/utils/security_config.py:
```diff
@@ -0,0 +1,132 @@
+            raise subprocess.CalledProcessError(proc.returncode, cmd)
+
+
+class SecurityConfig:
```
Same issue with new-style classes.

ewencp commented on tests/kafkatest/utils/security_config.py:
```diff
@@ -0,0 +1,132 @@
+
+    PLAINTEXT = 'PLAINTEXT'
+    SSL = 'SSL'
+    SSL_DIR = "/mnt/ssl"
```
I see a couple of these settings used elsewhere as literals rather than referencing these constants. We should replace them (grep for `/mnt/ssl`).

ewencp commented on tests/kafkatest/utils/security_config.py:
```diff
@@ -0,0 +1,132 @@
+
+        self.node_account = node_account
+        if security_protocol is None:
+            for line in template_props.split("\n"):
```
Is the unparsed form of the properties file what we really want here? It seems like a simple `dict` would be easier here. Is this just because we get this data out of rendered templates? If nothing else, perhaps we should factor out the parsing of the config file into a utility?

ewencp commented on tests/kafkatest/utils/security_config.py:
```diff
@@ -0,0 +1,132 @@
+        if security_protocol is None:
+            security_protocol = SecurityConfig.PLAINTEXT
+        if security_protocol == SecurityConfig.SSL:
+            node_account.ssh("mkdir -p %s" % SecurityConfig.SSL_DIR, allow_fail=False)
```
This class seems a bit weird -- it's constructor seems to be used like a function in a bunch of places (taking the node account and putting files in place, where you don't save the `SecurityConfig` object, just create it to get the side-effects of the constructor), but with a few extra helpers added in that you only occasionally hold on to the object for.

Should we split out the constructor, which ensures we have all the settings, from the part that actually configures the node?

Originally I was thinking that using a mixin, this would work naturally -- the determination of the configs would happen in `__init__`, then `start` or `start_node` would setup the files on the host.

ewencp commented on tests/kafkatest/utils/security_config.py:
```diff
@@ -0,0 +1,132 @@
+    
+    def write_to_file(self, prop_file='/mnt/ssl/security.properties'):
+        """Write security properties to a remote file"""
+        tmpfile =  tempfile.NamedTemporaryFile(delete=False)
```
There's a utility in ducktape on `RemoteAccount` for doing this already called `create_file`. Given a string with the contents, it'll generate the tempfile locally, scp it to the destination, and then clean up the tempfile.

ewencp commented:
@rajinisivaram This is looking good. I left a bunch of comments, but the core functionality that adds SSL test support seems solid. The changes I suggested are mostly targeted at code organization for the tests since the functionality seems ok.

rajinisivaram commented:
@ewencp @ijuma Thank you for review comments. I will work on the changes tomorrow and submit an update soon.

granders commented on tests/kafkatest/utils/security_config.py:
```diff
@@ -0,0 +1,132 @@
+        Generate JKS keystore and truststore and return
+        Kafka SSL properties with these stores.
+        """
+        ksPath = os.path.join(dir, 'test.keystore.jks')
```
Minor but important - change these camel-case variables to snake_case for consistency

rajinisivaram committed:
KAFKA-2581: Address review comments

rajinisivaram commented:
Test failure is unrelated to this PR.

Have addressed most of the issues raised. I didn't used mixin for security config, but have changed the code layout as suggested by @ewencp .

granders commented on tests/kafkatest/sanity_checks/test_console_consumer.py:
```diff
@@ -28,18 +30,23 @@ class ConsoleConsumerTest(Test):
     def __init__(self, test_context):
         super(ConsoleConsumerTest, self).__init__(test_context)
 
+        self.test_context = test_context
```
`self.test_context = test_context` is not necessary

every subclass of `Test` has this assigned as long as you make the proper call to `super(...).__init__(...)` (which you did)

granders commented on tests/kafkatest/services/console_consumer.py:
```diff
@@ -119,6 +123,22 @@ def __init__(self, context, num_nodes, kafka, topic, message_validator=None, fro
         self.message_validator = message_validator
         self.messages_consumed = {idx: [] for idx in range(1, num_nodes + 1)}
 
+        # Process client configuration
```
You can simply do
`self.prop_file = self.render('console_consumer.properties')`

The properties file already does the "null/None" check:

```
{% if consumer_timeout_ms is defined and consumer_timeout_ms is not none %}
consumer.timeout.ms={{ consumer_timeout_ms }}
{% endif %}
```

rajinisivaram replied:
@granders Have made these changes.

guozhangwang replied:
This seems not done in the latest patch.

rajinisivaram replied:
@guozhangwang I think the change suggested by @granders was to remove the if-then-else around render() which was in the code earlier. That was removed in the latest patch (the code selection above is slightly confusing). Can you let me know if I have misunderstood the comment?

granders replied:
@rajinisivaram @guozhangwang 
This was correctly updated - I was referring to the block around render()

granders commented on tests/kafkatest/utils/security_config.py:
```diff
@@ -0,0 +1,133 @@
+class Keytool(object):
+
+    @staticmethod
+    def generate_keystore_truststore(dir='.'):
```
`dir` is another reserved keyword, so another variable name is probably better

rajinisivaram committed:
KAFKA-2581: Remove unnecessary code

granders commented:
@rajinisivaram 
All tests passing: http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/98/console

granders commented on tests/kafkatest/services/templates/verifiable_producer.properties:
```diff
@@ -0,0 +1,16 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
I discussed this offline with @guozhang - to avoid having empty dummy files floating around, it may be better to delete this file and pass in empty string to in the `SecurityConfig` constructor in `verifiable_producer` (see comment below)

granders commented on tests/kafkatest/services/verifiable_producer.py:
```diff
@@ -37,7 +39,19 @@ def __init__(self, context, num_nodes, kafka, topic, max_messages=-1, throughput
 
+        # If security protocol is not specified, use the default in the template properties.
+        self.prop_file = self.render('verifiable_producer.properties')
+        self.security_config = SecurityConfig(security_protocol, self.prop_file)
```
As mentioned above, to avoid empty dummy files, we can just do something like this for now:

```
self.prop_file = ""
self.security_config = SecurityConfig(security_protocol, self.prop_file)
self.security_protocol = self.security_config.security_protocol
self.prop_file += str(self.security_config)
```

rajinisivaram replied:
@granders @guozhang The advantage of associating property files with all services like the verifiable producer is that it makes it possible to run producers with different default properties without modifying python code. We are planning to use this flexibility to run system tests with the security configuration we use (eg. endpoint validation on, client auth off, specific TLS version etc). If empty property files are an issue, would it be possible to change TemplateRenderer.render() in ducktape to treat missing property file as an empty file rather than remove support for property files in some services?

granders replied:
@rajinisivaram I think @guozhangwang has observed unnecessary empty stub files cluttering the code base in the past, and is suggesting that as a pattern to avoid.

Correct me if I'm wrong, but the way this logic is structured, it looks like like very little extra effort to add a default properties file as soon as non-empty defaults are needed (add the file, and switch to `self.prop_file = self.render(...)` Since this is such a minor edit, having an empty stub file in place doesn't really buy much.

As for rendering missing templates as empty strings in ducktape - I don't think this is the right approach, since it would hide error conditions and potentially cause confusing behavior. For example, if the user's intention is to use a nonempty template file, but the location is wrong, he or she should receive an error (easy to diagnose) than potentially start up the service with different settings than intended (harder to diagnose).

rajinisivaram replied:
@granders I do agree about the second part about rendering missing templates. I was thinking that even as I was typing the suggestion:-)

I am clearly not convincing enough about the first part. Let me try one last time. We want to run ducktape tests with different SSL parameters from the Kafka default. With the current code, we append some properties to verifiable_producer.properties and simply run ducktape. It gives us a test run that is not the same as the standard test run in Kafka. It is impractical to test every security property combination simply through paramterization of tests. The template properties file gives us the ability to run the tests in the one custom configuration that we are keen to test. Yes, it is straightforward to change python code to do the same too. But that is a code patch that we have to apply to every test run, not a property update. Patches need to be kept up-to-date with changes in verifiable_producer.py in a way property files dont. While the empty stub files look unnecessary, they are providing us the ability to run tests with a different configuration. If this flexibility is unlikely to be of value to anyone else, I am happy to remove the empty file and update the code.

guozhangwang replied:
@rajinisivaram Actually @granders and I are on the same page with you about using template producer property files to customize testing parameters moving forward. It is just that since we do not use it anywhere yet, adding a dummy empty file seems less attracting for now. So what we are suggesting is that:

a. Moving forward, we should 1) add an non-empty default property file, 2) always require a property file in verifiable_producer.py's config constrcution APIs, and add another API using default file if it is not specified.

b. For now, just keep the API that does not use any property files yet to override the property values, and also get rid of the dummy files.

Does this sound reasonable to you?

rajinisivaram replied:
@guozhangwang  OK, fixing the code now.

rajinisivaram committed:
KAFKA-2581: Remove properties file for verifiable_producer

guozhangwang commented:
LGTM.
